File: strands-ts/src/vended-tools/graph/graph.ts
/**
 * Graph tool: deterministic DAG orchestration over the SDK's Graph primitive.
 *
 * Shim over `Graph` in `@strands-agents/sdk/multiagent`. The caller agent
 * describes a DAG of sub-agents inline; this tool constructs a `Graph` from
 * those specs and invokes it, returning per-node text results.
 *
 * Design notes:
 * - **Shim only.** All orchestration lives in `multiagent/graph.ts`. This module
 *   validates the model-supplied topology, builds sub-agents, wires them into a
 *   `Graph`, and formats the result.
 * - **No edge conditions.** Conditional routing would require executing
 *   model-supplied code at every edge; not appropriate for a vended tool.
 * - **Tools are name-allow-listed** from the parent agent's registry. The model
 *   cannot conjure new tools inside a node — only reference tools the caller
 *   already has. Mirrors the `use_agent` / `swarm` convention.
 * - **Cycles are rejected** at validation time. `Graph` also requires reachability
 *   from source nodes; the tool detects cycles explicitly before construction so
 *   the failure message points at the actual problem.
 */

import { z } from 'zod'
import { Agent } from '../../agent/agent.js'
import { Graph } from '../../multiagent/graph.js'
import { AgentNode } from '../../multiagent/nodes.js'
import type { InvokableTool, Tool } from '../../tools/tool.js'
import { tool } from '../../tools/tool-factory.js'
import { TextBlock } from '../../types/messages.js'
import type { ContentBlock } from '../../types/messages.js'
import {
  DEFAULT_NODE_TIMEOUT_MS,
  DEFAULT_TIMEOUT_MS,
  GRAPH_DESCRIPTION,
  MAX_EDGES,
  MAX_ID_LENGTH,
  MAX_INITIAL_INPUT_LENGTH,
  MAX_NODES,
  MAX_STEPS,
  MAX_SYSTEM_PROMPT_LENGTH,
  MAX_TOOLS_PER_NODE,
  type GraphNodeResult,
  type GraphOutput,
} from './types.js'

/**
 * Shared multi-agent recursion depth counter — see `_multiagent-conventions.md`.
 * Every multi-agent vended tool (`use_agent`, `swarm`, `graph`, `a2a_client`)
 * reads/increments the same key so a chain crossing tool boundaries is still
 * capped.
 */
export const MULTIAGENT_DEPTH_KEY = 'multiagentDepth'
export const MAX_MULTIAGENT_DEPTH = 3

/**
 * Multi-agent tool names that must never appear in a child node's allow-list.
 * Depth-cap participation already blocks unbounded recursion, but the shared
 * convention (`_multiagent-conventions.md`) also mandates an explicit reject at
 * the tool boundary as defense-in-depth.
 */
const MULTIAGENT_TOOL_NAMES = new Set(['use_agent', 'swarm', 'graph', 'a2a_client'])

function currentDepth(invocationState: Record<string, unknown> | undefined): number {
  const raw = invocationState?.[MULTIAGENT_DEPTH_KEY]
  if (typeof raw !== 'number' || !Number.isInteger(raw) || raw < 0) return 0
  return raw
}

/**
 * Build a `cancelled` result. Spec (`_multiagent-conventions.md`): a cancelled
 * tool call returns `{status: 'cancelled'}` — it does not raise past the tool
 * boundary. A cancelled result reaching the loop is a signal to the parent,
 * not an exception to propagate.
 */
function cancelledOutput(startedAt: number, partial = ''): GraphOutput {
  return {
    status: 'cancelled',
    output: partial,
    executionOrder: [],
    results: {},
    executionTimeMs: Date.now() - startedAt,
  }
}

const nodeIdPattern = /^[A-Za-z0-9_-]+$/

const nodeSchema = z.object({
  id: z
    .string()
    .min(1, 'Node id must be non-empty.')
    .max(MAX_ID_LENGTH, `Node id must be at most ${MAX_ID_LENGTH} characters.`)
    .regex(nodeIdPattern, 'Node id must match [A-Za-z0-9_-].')
    .describe('Unique short identifier for the node.'),
  systemPrompt: z
    .string()
    .max(MAX_SYSTEM_PROMPT_LENGTH, `systemPrompt must be at most ${MAX_SYSTEM_PROMPT_LENGTH} characters.`)
    .optional()
    .describe("Optional system prompt for this node's sub-agent."),
  tools: z
    .array(z.string().min(1))
    .max(MAX_TOOLS_PER_NODE, `tools list must have at most ${MAX_TOOLS_PER_NODE} entries.`)
    .optional()
    .describe(
      "Tool names to expose to this node, drawn from the parent agent's tool " +
        'registry. Wildcards are not allowed; each tool must be listed by name.'
    ),
})

const edgeSchema = z.object({
  fromId: z.string().min(1).describe('Source node id.'),
  toId: z.string().min(1).describe('Target node id.'),
})

const graphInputSchema = z.object({
  nodes: z
    .array(nodeSchema)
    .min(1, "'nodes' must be a non-empty list.")
    .max(MAX_NODES, `Too many nodes (max ${MAX_NODES}).`)
    .describe('Nodes in the graph.'),
  edges: z
    .array(edgeSchema)
    .max(MAX_EDGES, `Too many edges (max ${MAX_EDGES}).`)
    .describe('Directed edges. Each entry declares that toId depends on fromId.'),
  initialInput: z
    .string()
    .max(MAX_INITIAL_INPUT_LENGTH, `initialInput must be at most ${MAX_INITIAL_INPUT_LENGTH} characters.`)
    .describe('The task passed to entry-point nodes (nodes with no incoming edges).'),
})

type NodeSpec = z.infer<typeof nodeSchema>
type EdgeSpec = z.infer<typeof edgeSchema>

/**
 * Verifies the graph has no cycles using Kahn's algorithm. Returns nothing on
 * success; throws when a cycle is detected. `Graph` is a DAG by design — cycle
 * detection at the tool boundary makes the failure message point at the
 * actual problem instead of a generic "unreachable node" error from the SDK.
 */
function assertNoCycle(nodes: NodeSpec[], edges: EdgeSpec[]): void {
  const ids = new Set(nodes.map((n) => n.id))
  const adjacency = new Map<string, string[]>()
  const inDegree = new Map<string, number>()
  for (const id of ids) inDegree.set(id, 0)

  for (const edge of edges) {
    if (!ids.has(edge.fromId)) {
      throw new Error(`Edge references unknown source node '${edge.fromId}'.`)
    }
    if (!ids.has(edge.toId)) {
      throw new Error(`Edge references unknown target node '${edge.toId}'.`)
    }
    if (edge.fromId === edge.toId) {
      throw new Error(`Self-loop on node '${edge.fromId}' is not allowed; the graph must be a DAG.`)
    }
    const targets = adjacency.get(edge.fromId) ?? []
    targets.push(edge.toId)
    adjacency.set(edge.fromId, targets)
    inDegree.set(edge.toId, (inDegree.get(edge.toId) ?? 0) + 1)
  }

  const queue: string[] = []
  for (const [id, deg] of inDegree.entries()) {
    if (deg === 0) queue.push(id)
  }
  let processed = 0
  while (queue.length > 0) {
    const current = queue.shift()!
    processed++
    for (const target of adjacency.get(current) ?? []) {
      const next = (inDegree.get(target) ?? 0) - 1
      inDegree.set(target, next)
      if (next === 0) queue.push(target)
    }
  }
  if (processed !== ids.size) {
    throw new Error('Graph contains a cycle; graphs must be acyclic (DAG).')
  }
}

/**
 * Assert unique ids in the flat `nodes` list. Zod validates each entry
 * individually but doesn't cross-check.
 */
function assertUniqueNodeIds(nodes: NodeSpec[]): void {
  const seen = new Set<string>()
  for (const node of nodes) {
    if (seen.has(node.id)) {
      throw new Error(`Duplicate node id '${node.id}'.`)
    }
    seen.add(node.id)
  }
}

/**
 * Resolve a node's tool allow-list against the parent registry. Rejects
 * wildcards, empty strings, and unknown names. Deduplicates.
 */
function resolveTools(names: string[] | undefined, parentTools: Map<string, Tool>, nodeId: string): Tool[] {
  if (!names || names.length === 0) return []
  const resolved: Tool[] = []
  const seen = new Set<string>()
  for (const raw of names) {
    if (raw === '*' || raw.includes('*')) {
      throw new Error(
        `Node '${nodeId}': tool '${raw}' looks like a wildcard. ` + 'List each tool by name; wildcards are not allowed.'
      )
    }
    // Reject multi-agent tools by name — defense-in-depth on top of the shared
    // depth cap. See `_multiagent-conventions.md`.
    if (MULTIAGENT_TOOL_NAMES.has(raw)) {
      throw new Error(
        `Node '${nodeId}': tool '${raw}' is a multi-agent tool and may not be ` +
          "used inside a graph node's allow-list."
      )
    }
    if (seen.has(raw)) continue
    seen.add(raw)
    const t = parentTools.get(raw)
    if (!t) {
      throw new Error(`Node '${nodeId}': tool '${raw}' is not registered on the parent agent.`)
    }
    resolved.push(t)
  }
  return resolved
}

/**
 * Coerce a node's ContentBlock output to a single text string. Non-text blocks
 * (images, tool uses, reasoning) are dropped — the tool's result shape is text
 * keyed by node id, which is what a downstream model will consume.
 */
function contentToText(blocks: ContentBlock[]): string {
  const parts: string[] = []
  for (const block of blocks) {
    if (block instanceof TextBlock) {
      parts.push(block.text)
    }
  }
  return parts.join('\n').trim()
}

/**
 * Options accepted by {@link makeGraph}. All fields are optional; each falls
 * back to the shared spec's default. Factory-only — none of these knobs are
 * surfaced to the model.
 */
export interface MakeGraphOptions {
  /** Tool name shown to the model. Defaults to `"graph"`. */
  name?: string
  /** Tool description shown to the model. */
  description?: string
  /** Shared multi-agent recursion cap. Defaults to {@link MAX_MULTIAGENT_DEPTH}. */
  maxDepth?: number
}

/**
 * Create a graph tool. Mirrors Python's `make_graph`. Use when you want a
 * different `name`, `description`, or `maxDepth` from the pre-built {@link graph}.
 *
 * @example
 * ```typescript
 * import { Agent } from '@strands-agents/sdk'
 * import { makeGraph } from '@strands-agents/sdk/vended-tools/graph'
 *
 * const shallowGraph = makeGraph({ maxDepth: 2 })
 * const agent = new Agent({ tools: [shallowGraph] })
 * ```
 */
export function makeGraph(
  options: MakeGraphOptions = {}
): InvokableTool<z.infer<typeof graphInputSchema>, GraphOutput> {
  const name = options.name ?? 'graph'
  const description = options.description ?? GRAPH_DESCRIPTION
  const maxDepth = options.maxDepth ?? MAX_MULTIAGENT_DEPTH

  return tool({
    name,
    description,
    inputSchema: graphInputSchema,
    callback: async (input, context): Promise<GraphOutput> => {
      if (!context) {
        throw new Error('graph tool requires a tool context.')
      }

      const invocationState = context.invocationState as Record<string, unknown> | undefined
      const depth = currentDepth(invocationState)
      if (depth >= maxDepth) {
        throw new Error(
          `graph refused: multi-agent recursion depth cap of ${maxDepth} reached (current depth ${depth})`
        )
      }

      const { nodes: nodeSpecs, edges: edgeSpecs, initialInput } = input

      assertUniqueNodeIds(nodeSpecs)
      assertNoCycle(nodeSpecs, edgeSpecs)

      const startedAt = Date.now()

      if (context.agent.cancelSignal.aborted) {
        // Pre-flight cancellation: return the cancelled sentinel instead of
        // raising. Spec (`_multiagent-conventions.md`): the tool returns
        // `{status: 'cancelled'}`, does not raise past the tool boundary.
        return cancelledOutput(startedAt)
      }

      const parentToolMap = new Map<string, Tool>()
      for (const t of context.agent.toolRegistry.list()) {
        parentToolMap.set(t.name, t)
      }

      const parentModel = context.agent.model

      const agentNodes: AgentNode[] = nodeSpecs.map((spec) => {
        const nodeTools = resolveTools(spec.tools, parentToolMap, spec.id)
        const subAgent = new Agent({
          id: spec.id,
          model: parentModel,
          printer: false,
          ...(spec.systemPrompt !== undefined && { systemPrompt: spec.systemPrompt }),
          tools: nodeTools,
        })
        return new AgentNode({ agent: subAgent })
      })

      const g = new Graph({
        nodes: agentNodes,
        edges: edgeSpecs.map((e) => ({ source: e.fromId, target: e.toId })),
        timeout: DEFAULT_TIMEOUT_MS,
        nodeTimeout: DEFAULT_NODE_TIMEOUT_MS,
        maxSteps: MAX_STEPS,
      })

      const childInvocationState: Record<string, unknown> = { [MULTIAGENT_DEPTH_KEY]: depth + 1 }

      // Hard wall-clock ceiling. The SDK `Graph`'s own `timeout` is checked at
      // node boundaries, so an N-node chain of long-running nodes can silently
      // exceed the cap. Composing an `AbortSignal.timeout` with the parent's
      // cancel signal gives a strict upper bound that fires regardless of where
      // execution is inside the graph — parity with the Python side's
      // `asyncio.wait_for` guard.
      const hardCeiling = AbortSignal.timeout(DEFAULT_TIMEOUT_MS)
      const composedSignal = AbortSignal.any([context.agent.cancelSignal, hardCeiling])

      let result
      try {
        result = await g.invoke(initialInput, {
          cancelSignal: composedSignal,
          invocationState: childInvocationState,
        })
      } catch (err) {
        // The SDK's `Graph` throws when its `cancelSignal` aborts mid-flight
        // (see `graph.ts:343`). Map that back to a cancelled result — spec
        // forbids raising past the tool boundary. Anything else is a real
        // error and should propagate. The hard-ceiling `AbortSignal.timeout`
        // is treated the same way: no useful result once the cap fires.
        if (composedSignal.aborted) {
          return cancelledOutput(startedAt)
        }
        throw err
      }

      const results: Record<string, GraphNodeResult> = {}
      const executionOrder: string[] = []
      for (const nodeResult of result.results) {
        results[nodeResult.nodeId] = {
          // Lowercase the SDK enum value so the tool result matches the Python
          // side and the shared multi-agent dialect (see
          // `_multiagent-conventions.md`), which specifies single-word lowercase
          // status strings.
          status: nodeResult.status.toLowerCase(),
          output: nodeResult.error ? `error: ${nodeResult.error.message}` : contentToText(nodeResult.content),
          executionTimeMs: nodeResult.duration,
        }
        executionOrder.push(nodeResult.nodeId)
      }

      const terminalOutput = aggregateTerminalOutput(nodeSpecs, edgeSpecs, results)

      return {
        status: result.status.toLowerCase(),
        output: terminalOutput,
        executionOrder,
        results,
        executionTimeMs: result.duration,
      }
    },
  })
}

/**
 * Pre-built graph tool. Register on an `Agent` alongside any tools you want
 * the graph's nodes to be able to call.
 *
 * @example
 * ```typescript
 * import { Agent } from '@strands-agents/sdk'
 * import { graph } from '@strands-agents/sdk/vended-tools/graph'
 *
 * const agent = new Agent({ tools: [graph] })
 * ```
 */
export const graph = makeGraph()

/**
 * Concatenate every terminal (leaf) node's output text, joined by a blank
 * line, in graph declaration order. A graph without a single named sink
 * typically returns multiple final outputs; dropping any of them would
 * silently truncate the tool's answer.
 */
function aggregateTerminalOutput(
  nodes: NodeSpec[],
  edges: EdgeSpec[],
  results: Record<string, GraphNodeResult>
): string {
  const withOutgoing = new Set<string>(edges.map((e) => e.fromId))
  const parts: string[] = []
  for (const spec of nodes) {
    if (withOutgoing.has(spec.id)) continue
    const nodeResult = results[spec.id]
    if (!nodeResult) continue
    if (nodeResult.output) parts.push(nodeResult.output)
  }
  return parts.join('\n\n')
}

File: strands-py/src/strands/vended_tools/graph/graph.py
"""Graph tool: deterministic DAG orchestration over the SDK's Graph primitive.

Shim over :class:`~strands.multiagent.graph.GraphBuilder` /
:class:`~strands.multiagent.graph.Graph`. The caller agent describes a DAG of
sub-agents inline; this tool constructs the ``Graph`` and invokes it, returning
per-node text results.

Design notes:
    * The tool is a **shim**. All orchestration lives in ``strands.multiagent.graph``.
      This module only validates the model-supplied topology, builds sub-agents,
      wires them into a ``Graph``, and formats the result.
    * **No edge conditions**. Conditional routing would require executing
      model-supplied code at every edge; that is not appropriate for a vended tool.
      Users who want conditional routing should build a ``Graph`` directly.
    * **Tools are name-allow-listed** from the parent agent's registry. The model
      cannot conjure new tools inside a node — only reference tools the caller
      already has. This mirrors the ``use_agent`` / ``swarm`` convention.
    * **Cycles are rejected** at validation time. ``Graph`` supports cycles;
      the tool does not.
    * **Recursion depth** participates in the shared ``multiagent_depth`` counter
      threaded through ``invocation_state``. See ``_multiagent_conventions.md``.
"""

from __future__ import annotations

import asyncio
import time
from collections import defaultdict, deque
from typing import TYPE_CHECKING, Any, cast

from ...agent import Agent
from ...hooks import BeforeNodeCallEvent
from ...multiagent.base import Status
from ...multiagent.graph import GraphBuilder
from ...tools.decorator import tool
from ...types.tools import ToolContext
from .types import (
    DEFAULT_EXECUTION_TIMEOUT_SECONDS,
    DEFAULT_NODE_TIMEOUT_SECONDS,
    GRAPH_DESCRIPTION,
    MAX_EDGES,
    MAX_ID_LENGTH,
    MAX_INITIAL_INPUT_LENGTH,
    MAX_NODE_EXECUTIONS,
    MAX_NODES,
    MAX_SYSTEM_PROMPT_LENGTH,
    MAX_TOOLS_PER_NODE,
    GraphNodeResult,
    GraphOutput,
)

# Shared multi-agent depth counter — see ``_multiagent_conventions.md``. All
# multi-agent vended tools (`use_agent`, `swarm`, `graph`, `a2a_client`)
# increment the same key so a chain that crosses tool boundaries is still
# capped.
_DEPTH_KEY = "multiagent_depth"
_DEFAULT_MAX_DEPTH = 3

# Multi-agent tool names that must never appear in a child node's allow-list.
# Depth-cap participation already blocks unbounded recursion, but the shared
# convention (``_multiagent_conventions.md``) also mandates an explicit reject
# at the tool boundary as defense-in-depth — a compromised parent registry could
# expose these under aliased names, but a reject on the canonical name still
# catches the common case.
_MULTIAGENT_TOOL_NAMES = frozenset({"use_agent", "swarm", "graph", "a2a_client"})

if TYPE_CHECKING:
    from ...tools.decorator import DecoratedFunctionTool
    from ...types.tools import AgentTool


class MultiagentDepthExceeded(RuntimeError):
    """Raised when the shared multi-agent recursion cap is reached."""


# Sentinel embedded in the ``cancel_node`` message. The SDK's ``Graph`` re-raises
# that message as a ``RuntimeError``, so we need to distinguish "cancelled by
# our hook" from "some other runtime failure that mentioned cancellation".
_CANCEL_SENTINEL = "graph-tool-parent-cancel"
_CANCEL_MESSAGE = f"cancelled by parent agent [{_CANCEL_SENTINEL}]"


class _ParentCancelHook:
    """Cancel each graph node when the parent agent's cancel signal is set.

    The SDK exposes ``BeforeNodeCallEvent.cancel_node``; setting that field
    stops execution cleanly at the next node boundary. Polling the parent's
    ``threading.Event`` inside a hook keeps propagation on the same task as
    the graph itself, which is the only correct place for it — the previous
    finally-block approach only ran after ``invoke_async`` had already returned.
    """

    def __init__(self, parent_agent: Any) -> None:
        self._parent = parent_agent

    def on_before_node_call(self, event: BeforeNodeCallEvent) -> None:
        cancel_signal = getattr(self._parent, "_cancel_signal", None)
        if cancel_signal is not None and cancel_signal.is_set():
            event.cancel_node = _CANCEL_MESSAGE


def _current_depth(invocation_state: dict[str, Any]) -> int:
    raw = invocation_state.get(_DEPTH_KEY, 0)
    # ``bool`` is a subclass of ``int``; reject it explicitly so ``True`` doesn't
    # silently count as depth 1. Not exploitable, but a hostile ``invocation_state``
    # shouldn't be able to confuse the counter.
    if isinstance(raw, bool) or not isinstance(raw, int) or raw < 0:
        return 0
    return int(raw)


def _cancelled_output(execution_time_ms: int, partial: str = "") -> GraphOutput:
    """Build a ``cancelled`` result.

    Spec (see ``_multiagent_conventions.md``) forbids raising past the tool
    boundary; a cancelled result reaching the loop is a signal to the parent,
    not an exception to propagate.
    """
    return GraphOutput(
        status="cancelled",
        output=partial,
        execution_order=[],
        results={},
        execution_time_ms=execution_time_ms,
    )


def _validate_node_id(node_id: Any) -> str:
    if not isinstance(node_id, str) or not node_id:
        raise ValueError("Each node must have a non-empty string 'id'.")
    if len(node_id) > MAX_ID_LENGTH:
        raise ValueError(f"Node id '{node_id[:16]}...' exceeds {MAX_ID_LENGTH} characters.")
    if not all(ch.isalnum() or ch in "_-" for ch in node_id):
        raise ValueError(
            f"Node id '{node_id}' contains characters outside [A-Za-z0-9_-]. "
            "Use short identifiers for readability and to avoid collisions."
        )
    return node_id


def _resolve_tools(names: list[Any] | None, parent_registry: dict[str, AgentTool], node_id: str) -> list[AgentTool]:
    """Resolve a node's tool allow-list from the parent registry.

    Args:
        names: List of tool names the model asked for on this node. ``None`` or
            an empty list means the node runs with no tools.
        parent_registry: The caller agent's tool registry (name -> tool).
        node_id: Node id, for error messages.

    Raises:
        ValueError: If ``names`` is not a list of strings, contains a wildcard,
            or references a tool the parent registry does not expose.
    """
    if names is None:
        return []
    if not isinstance(names, list):
        raise ValueError(f"Node '{node_id}': 'tools' must be a list of tool names.")
    if len(names) > MAX_TOOLS_PER_NODE:
        raise ValueError(
            f"Node '{node_id}': 'tools' has {len(names)} entries (max {MAX_TOOLS_PER_NODE})."
        )

    resolved: list[AgentTool] = []
    seen: set[str] = set()
    for raw in names:
        if not isinstance(raw, str) or not raw:
            raise ValueError(f"Node '{node_id}': tool entries must be non-empty strings.")
        # Reject wildcards explicitly — a name-allow-list, never "all parent tools".
        if raw == "*" or "*" in raw:
            raise ValueError(
                f"Node '{node_id}': tool '{raw}' looks like a wildcard. "
                "List each tool by name; wildcards are not allowed."
            )
        # Reject multi-agent tools by name — defense-in-depth on top of the
        # shared depth cap. See ``_multiagent_conventions.md``.
        if raw in _MULTIAGENT_TOOL_NAMES:
            raise ValueError(
                f"Node '{node_id}': tool '{raw}' is a multi-agent tool and may not be "
                "used inside a graph node's allow-list."
            )
        if raw in seen:
            continue
        seen.add(raw)
        tool_instance = parent_registry.get(raw)
        if tool_instance is None:
            raise ValueError(f"Node '{node_id}': tool '{raw}' is not registered on the parent agent.")
        resolved.append(tool_instance)
    return resolved


def _validate_nodes_edges(nodes: list[dict[str, Any]], edges: list[dict[str, Any]]) -> None:
    """Structural checks that must pass before we build any sub-agents.

    Verifies node/edge counts, unique ids, that edges reference known nodes,
    and that the resulting graph has no cycles. The SDK ``GraphBuilder`` also
    accepts cycles (they're valid for cyclic multi-agent patterns), so cycle
    detection is the tool's responsibility.
    """
    if not isinstance(nodes, list) or not nodes:
        raise ValueError("'nodes' must be a non-empty list.")
    if len(nodes) > MAX_NODES:
        raise ValueError(f"Too many nodes: {len(nodes)} (max {MAX_NODES}).")

    if not isinstance(edges, list):
        raise ValueError("'edges' must be a list (may be empty).")
    if len(edges) > MAX_EDGES:
        raise ValueError(f"Too many edges: {len(edges)} (max {MAX_EDGES}).")

    ids: set[str] = set()
    for node in nodes:
        if not isinstance(node, dict):
            raise ValueError("Each node must be an object.")
        node_id = _validate_node_id(node.get("id"))
        if node_id in ids:
            raise ValueError(f"Duplicate node id '{node_id}'.")
        ids.add(node_id)

    adjacency: dict[str, list[str]] = defaultdict(list)
    indegree: dict[str, int] = {node_id: 0 for node_id in ids}
    for edge in edges:
        if not isinstance(edge, dict):
            raise ValueError("Each edge must be an object with 'from_id' and 'to_id'.")
        from_id = edge.get("from_id")
        to_id = edge.get("to_id")
        if not isinstance(from_id, str) or not isinstance(to_id, str):
            raise ValueError("Edge 'from_id' and 'to_id' must be strings.")
        if from_id not in ids:
            raise ValueError(f"Edge references unknown source node '{from_id}'.")
        if to_id not in ids:
            raise ValueError(f"Edge references unknown target node '{to_id}'.")
        if from_id == to_id:
            raise ValueError(f"Self-loop on node '{from_id}' is not allowed; the graph must be a DAG.")
        adjacency[from_id].append(to_id)
        indegree[to_id] += 1

    # Kahn's algorithm — if we can't peel all nodes off, the remainder is a cycle.
    queue = deque([node_id for node_id, deg in indegree.items() if deg == 0])
    processed = 0
    while queue:
        current = queue.popleft()
        processed += 1
        for neighbor in adjacency[current]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                queue.append(neighbor)
    if processed != len(ids):
        raise ValueError("Graph contains a cycle; graphs must be acyclic (DAG).")


def _build_agent_for_node(
    node: dict[str, Any],
    parent_agent: Any,
    parent_registry: dict[str, AgentTool],
) -> Agent:
    """Build one sub-agent for a graph node.

    * ``model`` defaults to the parent agent's model — this keeps sub-agents on a
      known, working provider without letting the model pick arbitrary strings.
    * ``system_prompt`` is optional and length-capped.
    * ``tools`` is a strict allow-list against the parent's registry.
    """
    node_id = node["id"]

    system_prompt = node.get("system_prompt")
    if system_prompt is not None:
        if not isinstance(system_prompt, str):
            raise ValueError(f"Node '{node_id}': 'system_prompt' must be a string.")
        if len(system_prompt) > MAX_SYSTEM_PROMPT_LENGTH:
            raise ValueError(f"Node '{node_id}': 'system_prompt' exceeds {MAX_SYSTEM_PROMPT_LENGTH} characters.")

    tools = _resolve_tools(node.get("tools"), parent_registry, node_id)

    return Agent(
        model=parent_agent.model,
        name=node_id,
        system_prompt=system_prompt,
        tools=list(tools),
    )


def _format_result(result: Any) -> tuple[str, int]:
    """Turn a NodeResult into ``(text, execution_time_ms)``.

    Extracts text blocks explicitly from the underlying ``AgentResult`` rather
    than delegating to ``AgentResult.__str__``, which stringifies interrupts
    and structured output ahead of the text content. Explicit extraction keeps
    the tool byte-parity with the TypeScript side's ``contentToText`` and
    means a future change to ``__str__`` cannot silently alter tool output.

    Falls back to a short error description when the node failed. The wording
    matches the TypeScript side ("error: <message>") so a caller consuming
    either SDK sees byte-identical output for the same failure.
    """
    execution_time = int(getattr(result, "execution_time", 0) or 0)
    inner = getattr(result, "result", None)

    if isinstance(inner, Exception):
        return f"error: {inner}", execution_time

    if inner is None:
        return "", execution_time

    # ``AgentResult`` exposes ``.message`` (a ``Message`` dict with ``content``,
    # a list of content blocks). Extract only text blocks; images, tool uses,
    # reasoning, and other structured blocks are dropped so the tool result
    # cannot smuggle non-text bytes back through the graph.
    message = getattr(inner, "message", None)
    if isinstance(message, dict):
        content_blocks = message.get("content", [])
        parts: list[str] = []
        if isinstance(content_blocks, list):
            for block in content_blocks:
                if isinstance(block, dict) and isinstance(block.get("text"), str):
                    parts.append(block["text"])
        return "\n".join(parts).strip(), execution_time

    # Nested ``MultiAgentResult`` or an unrecognised shape: fall back to str().
    return str(inner), execution_time


def make_graph(
    *,
    name: str = "graph",
    description: str = GRAPH_DESCRIPTION,
    max_depth: int = _DEFAULT_MAX_DEPTH,
) -> DecoratedFunctionTool:
    """Create a graph tool bound to no specific parent.

    Args:
        name: Tool name shown to the model. Defaults to ``"graph"``.
        description: Tool description shown to the model.
        max_depth: Shared multi-agent recursion cap. Factory-only; never
            surfaced to the model.

    Returns:
        A decorated tool. Register it on an :class:`~strands.agent.Agent` like
        any other tool.
    """

    @tool(name=name, description=description, context="tool_context")
    async def graph_tool(
        nodes: list[dict[str, Any]],
        edges: list[dict[str, Any]],
        initial_input: str,
        tool_context: ToolContext,
    ) -> GraphOutput:
        """Execute a directed acyclic graph of sub-agents.

        Args:
            nodes: List of node specs. Each entry has:

                * ``id`` (required): unique short identifier (``[A-Za-z0-9_-]``, up to 64 chars).
                * ``system_prompt`` (optional): sub-agent system prompt.
                * ``tools`` (optional): list of tool names to expose to this node,
                  drawn from the parent agent's tool registry. No wildcards.
            edges: List of ``{"from_id": ..., "to_id": ...}`` entries defining the
                dependency order. The graph must be acyclic.
            initial_input: The task passed to entry-point nodes (nodes with no
                incoming edges).
            tool_context: Injected by the framework. Not user-facing.

        Raises:
            ValueError: If the topology is invalid (non-list ``nodes`` / ``edges``,
                duplicate or malformed node ids, self-loops, cycles, references to
                unknown nodes), if ``initial_input`` is not a string or exceeds the
                size cap, if a node's ``system_prompt`` exceeds the size cap, or if
                its ``tools`` field references an unknown tool, a wildcard, a
                multi-agent tool name, or exceeds the per-node tool cap.
            MultiagentDepthExceeded: If the shared multi-agent recursion depth
                counter (``invocation_state['multiagent_depth']``) is at or above
                the factory-configured cap on entry.
        """
        # Validation (bad-input) errors still raise past the tool boundary —
        # they signal a bug the model should see. Only cancellation is caught
        # and mapped to a ``{status: cancelled}`` result per spec.
        invocation_state = tool_context.invocation_state or {}
        depth = _current_depth(invocation_state)
        if depth >= max_depth:
            raise MultiagentDepthExceeded(
                f"graph refused: multi-agent recursion depth cap of {max_depth} reached (current depth {depth})"
            )

        if not isinstance(initial_input, str):
            raise ValueError("'initial_input' must be a string.")
        if len(initial_input) > MAX_INITIAL_INPUT_LENGTH:
            raise ValueError(f"'initial_input' exceeds {MAX_INITIAL_INPUT_LENGTH} characters.")

        _validate_nodes_edges(nodes, edges)

        parent_agent = tool_context.agent
        parent_registry_container = getattr(parent_agent, "tool_registry", None)
        parent_registry: dict[str, AgentTool] = (
            getattr(parent_registry_container, "registry", {}) if parent_registry_container else {}
        )

        parent_cancel_signal = getattr(parent_agent, "_cancel_signal", None)
        start_time = time.monotonic()
        if parent_cancel_signal is not None and parent_cancel_signal.is_set():
            # Pre-flight cancellation: return the cancelled sentinel instead of
            # raising. Spec (``_multiagent_conventions.md`` line 80): the tool
            # returns ``{status: cancelled}``, does not raise.
            return _cancelled_output(execution_time_ms=int((time.monotonic() - start_time) * 1000))

        builder = GraphBuilder()
        builder.set_max_node_executions(MAX_NODE_EXECUTIONS)
        builder.set_execution_timeout(DEFAULT_EXECUTION_TIMEOUT_SECONDS)
        builder.set_node_timeout(DEFAULT_NODE_TIMEOUT_SECONDS)

        for node in nodes:
            sub_agent = _build_agent_for_node(node, parent_agent, parent_registry)
            builder.add_node(sub_agent, node_id=node["id"])

        for edge in edges:
            builder.add_edge(edge["from_id"], edge["to_id"])

        graph_instance = builder.build()
        graph_instance.add_hook(_ParentCancelHook(parent_agent).on_before_node_call, BeforeNodeCallEvent)

        child_invocation_state: dict[str, Any] = {_DEPTH_KEY: depth + 1}

        # Hard wall-clock ceiling. ``builder.set_execution_timeout`` is checked
        # at node boundaries, so an N-node chain can silently exceed the cap.
        # ``asyncio.wait_for`` gives us a strict upper bound at exactly the
        # documented cap so the tool's guarantee matches its documentation.
        hard_ceiling = DEFAULT_EXECUTION_TIMEOUT_SECONDS

        try:
            result = await asyncio.wait_for(
                graph_instance.invoke_async(initial_input, invocation_state=child_invocation_state),
                timeout=hard_ceiling,
            )
        except (asyncio.TimeoutError, TimeoutError):
            # Hard-ceiling exceeded. Surface as cancelled — the tool boundary
            # doesn't distinguish "cancelled by parent" from "cancelled by the
            # tool's own hard timeout"; both mean "no useful result".
            return _cancelled_output(execution_time_ms=int((time.monotonic() - start_time) * 1000))
        except RuntimeError as exc:
            # ``Graph.invoke_async`` re-raises ``RuntimeError`` when a hook sets
            # ``event.cancel_node``. Match on the sentinel embedded in the hook's
            # message so an unrelated runtime failure that happens to coincide
            # with parent cancellation still propagates rather than being
            # silently mapped to a cancelled result.
            if _CANCEL_SENTINEL not in str(exc):
                raise
            return _cancelled_output(execution_time_ms=int((time.monotonic() - start_time) * 1000))

        results: dict[str, GraphNodeResult] = {}
        for node_id, node_result in result.results.items():
            text, execution_time_ms = _format_result(node_result)
            results[node_id] = GraphNodeResult(
                status=node_result.status.value,
                output=text,
                execution_time_ms=execution_time_ms,
            )

        top_output = _aggregate_terminal_output(nodes, edges, results)

        return GraphOutput(
            status=result.status.value if isinstance(result.status, Status) else str(result.status),
            output=top_output,
            execution_order=[node.node_id for node in result.execution_order],
            results=results,
            execution_time_ms=int(result.execution_time),
        )

    return cast("DecoratedFunctionTool", graph_tool)


def _aggregate_terminal_output(
    nodes: list[dict[str, Any]],
    edges: list[dict[str, Any]],
    results: dict[str, GraphNodeResult],
) -> str:
    """Return the top-level ``output`` string per the shared multi-agent dialect.

    Rule: concatenate every terminal (leaf) node's ``output`` text, joined by a
    blank line, in graph declaration order. Terminal nodes are those with no
    outgoing edges — a graph without a single named sink typically returns
    multiple final outputs, and dropping any of them would silently truncate
    the tool's answer.
    """
    with_outgoing: set[str] = {edge["from_id"] for edge in edges if isinstance(edge.get("from_id"), str)}
    parts: list[str] = []
    for node in nodes:
        node_id = node["id"]
        if node_id in with_outgoing:
            continue
        node_result = results.get(node_id)
        if node_result is None:
            continue
        text = node_result["output"]
        if text:
            parts.append(text)
    return "\n\n".join(parts)


graph = make_graph()
